monai/metrics/rocauc.py

```diff
@@ -18,33 +18,36 @@
 def _calculate(y, y_pred):
     assert y.ndimension() == y_pred.ndimension() == 1 and len(y) == len(y_pred), \
         'y and y_pred must be 1 dimension data with same length.'
-    assert y.unique().equal(torch.tensor([0, 1])), 'y values must be 0 or 1, can not be all 0 or all 1.'
-    pos_indexes, neg_indexes = list(), list()
-    for index, y_ in enumerate(y):
-        if y_ == 1:
-            pos_indexes.append(index)
-        elif y_ == 0:
-            neg_indexes.append(index)
+    assert y.unique().equal(torch.tensor([0, 1], device=y.device)), \
+        'y values must be 0 or 1, can not be all 0 or all 1.'
+    n = len(y)
+    indexes = y_pred.argsort()
+    y = y[indexes].cpu().numpy()
+    y_pred = y_pred[indexes].cpu().numpy()
+    nneg = auc = tmp_pos = tmp_neg = 0
+
+    for i in range(n):
+        y_i = y[i]
+        if i + 1 < n and y_pred[i] == y_pred[i + 1]:
+            tmp_pos += y_i
+            tmp_neg += 1 - y_i
+            continue
+        if tmp_pos + tmp_neg > 0:
+            tmp_pos += y_i
+            tmp_neg += 1 - y_i
+            nneg += tmp_neg
+            auc += tmp_pos * (nneg - tmp_neg / 2)
+            tmp_pos = tmp_neg = 0
+            continue
+        if y_i == 1:
+            auc += nneg
         else:
-            raise ValueError('labels should be binary values.')
-
-    def compare(a, b):
-        if a > b:
-            return 1
-        elif a == b:
-            return 0.5
-        else:
-            return 0
-    auc = sum([sum([compare(y_pred[i], y_pred[j]) for j in neg_indexes]) for i in pos_indexes])
-
-    return auc / (len(pos_indexes) * len(neg_indexes))
+            nneg += 1
+    return auc / (nneg * (n - nneg))
 
 
 def compute_roc_auc(y_pred, y, to_onehot_y=False, add_softmax=False, add_sigmoid=False, average='macro'):
-    """Computes Area Under the Receiver Operating Characteristic Curve (ROC AUC) based on:
-    `sklearn.metrics.roc_auc_score <http://scikit-learn.org/stable/modules/generated/
-    sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score>`_ .
-
+    """Computes Area Under the Receiver Operating Characteristic Curve (ROC AUC).
```

Comment:
I think we still need the link to sklearn rocauc, because we're following their convention for the `average` type.  Also perhaps we don't need the `add_sigmoid` argument here, because True/False wouldn't change the auc value. apart from these the code looks fine to me.

Reply:
updated in the latest patch.
Thanks.